WordPress/gutenberg
# [Mobile] List V2 Support

**Main PRs:**
- `Gutenberg Mobile` -> https://github.com/wordpress-mobile/gutenberg-mobile/pull/5054
- `WordPress iOS` -> https://github.com/wordpress-mobile/WordPress-iOS/pull/19116
- `WordPress Android` -> https://github.com/wordpress-mobile/WordPress-Android/pull/16962

Fixes https://github.com/WordPress/gutenberg/issues/42624

**Related PRs:**

- [x] https://github.com/WordPress/gutenberg/pull/42697
- [x] https://github.com/WordPress/gutenberg/pull/42703

## What?
This PR adds support for the new List V2 block on mobile, including enabling this new version using the `__experimentalEnableListBlockV2` feature flag that will be fetched through the block editor settings endpoint.

## Why?
To keep compatibility with the new changes in the web editor and prevent issues when rendering a block format that is not supported.

## How?

Adds compatibility sharing the same code with Web for the [edit.js](https://github.com/WordPress/gutenberg/pull/42702/files#diff-8ca208404bc35d0549ee68bfdb5c87a0d1fa101ddcfad1095d9a3c0e171b0272) file, with some required changes like platform-specific TagName and styles.

It adds the new `ListItem` component, this one has its own native variant since it required quite a few changes making it harder to share with Web, although it does uses shared code like the `IndentUI` component.

It also introduces a new component `ListStyleType` this was reviewed in a separate PR to make it easier to review, you can see just those changes in [this PR](https://github.com/WordPress/gutenberg/pull/42703). This new component will take care of the list styling, whether if its an ordered list or a standard one with their different indentation levels.

Introduces the `__experimentalEnableListBlockV2` feature flag logic as we did for the Quote V2 block, this will be in the code base for a brief time so we could make a mobile release with the new format disabled and later on activate it once the Web has it available through the Gutenberg plugin. In a following mobile release, this feature flag logic will be removed and List V2 block will be enabled by default.

Lastly, it introduces some Global styles changes we did in [this PR](https://github.com/WordPress/gutenberg/pull/42703), to parse font size values that were in a different format, like `rem` values. 

Unit tests have been added for the new List V2 format and kept two simple ones for the older version.

## Testing Instructions

**Precondition:** A site with the Gutenberg 13.8 plugin.

- Use a self-hosted site and install the latest Gutenberg plugin version >= 13.8.0. You can install the latest version [here](https://github.com/WordPress/gutenberg/releases).
- Once the latest version is installed, go to your site's admin page > Gutenberg > Experiments.
- Enable List block v2.

**NOTE:** Once the flag is enabled sometimes the new List block won't be shown the first time the editor is opened, so if you still see the old version without inner blocks, close the editor and open it again.

### Test case 1 - Test different levels of indentation

- Open the app using the builds from the main apps
- Add a List block
- Add a few items
- Change the indentation for some of them
- **Expect** to see the list update accordingly and show the expected icons depending on the level.
- Try merging different items or creating new ones

### Test case 2 - Test different levels of indentation and ordered lists

- Open the app using the builds from the main apps
- Add a List block
- Add a few items
- Change the indentation for some of them
- **Expect** to see the list update accordingly and show the expected icons depending on the level.
- Try merging different items or creating new ones
- Change the list settings to be an ordered one
- **Expect** to see the numbers instead of the icons
- Change the order using the `reverse` option
- **Expect** to see the numbers reversed.
- Set a starting value for the list
- **Expect** the list to update accordingly

### Test case 3 - Set a different font size

- Open the app using the builds from the main apps
- Add a List block
- Add a few items
- Open the block's settings
- Change the font size
- **Expect** the list's font size to be updated

### Test case 4 - Set some background and text colors

- Open the app using the builds from the main apps
- Add a List block
- Add a few items
- Open the block's settings
- Change the background and text color for the list
- **Expect** to see the new colors applied
- Try setting color customization for nested lists

## Screenshots or screencast 

### Test case 1 - Test different levels of indentation

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/181581938-c94d0157-aa07-4076-bfd5-c9a69a82661f.png" width="200" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/181581970-83d945a6-133e-4b56-b2be-6f0ea59a460d.png" width="200" /></kbd>

### Test case 2 - Test different levels of indentation and ordered lists

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/181582361-31a86282-b776-4038-9795-1ce0bc2e0c52.png" width="200" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/181582380-af55084e-05d1-47f9-bea1-3a54b8830bc5.png" width="200" /></kbd>

### Test case 3 - Set a different font size

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/181583208-a8ad9f0e-3b14-460c-a8e1-f2e8c291aefe.png" width="200" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/181583254-c218954a-139a-432f-874e-d0fa0b14142e.png" width="200" /></kbd>

### Test case 4 - Set some background and text colors

iOS | Android
-|-
<kbd><img src="https://user-images.githubusercontent.com/4885740/181594170-cbe218a9-c72c-46b1-979a-0293bbd06523.png" width="200" /></kbd> | <kbd><img src="https://user-images.githubusercontent.com/4885740/181594184-d4e0ec08-d635-451c-960c-1d39acad5efe.png" width="200" /></kbd>
